polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/TokenBucketRateLimiter.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.ratelimiter;
+
+/**
+ * Token bucket implementation of a Polaris RateLimiter. Acquires tokens at a fixed rate and has a
+ * maximum amount of tokens. Each "acquire" costs 1 token.
+ */
+public class TokenBucketRateLimiter implements RateLimiter {
```

Comment:
This seems like it could be a `Semaphore`

Reply:
I think a Semaphore would work as a limit on the number of concurrent requests that blocks instead of rejecting the request. This is a limit on request rate that rejects requests.

Reply:
Java's Semaphore has a [tryAcquire](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Semaphore.html#tryAcquire()), and you could reject the request if that returns `false`.

Reply:
Ah I see, then it would work if the intent was a concurrent request limit. I think there's room for debate on whether or not there should be a concurrent request limiter but it doesn't feel like that has to be answered in this PR.